File: Fastmate/MainWindowController.m
#import "MainWindowController.h"
#import "WebViewController.h"
#import "KVOBlockObserver.h"
#import "UserDefaultsKeys.h"
#import "Fastmate-Swift.h"

@interface MainWindowController () <NSWindowDelegate>

@property (nonatomic, strong) id titleBarSettingObserver;
@property (nonatomic, strong) id titleObserver;

@end

@implementation MainWindowController

- (void)windowDidLoad {
    [super windowDidLoad];

    // Fixes that we can't trust that the main window exists in applicationDidFinishLaunching:.
    // Here we always know that this content view controller will be the main web view controller,
    // so inform the app delegate
    AppDelegate *appDelegate = (AppDelegate *)NSApplication.sharedApplication.delegate;
    appDelegate.mainWebViewController = (WebViewController *)self.contentViewController;

    NSColor *windowColor = [NSKeyedUnarchiver unarchiveObjectWithData:[NSUserDefaults.standardUserDefaults dataForKey:WindowBackgroundColorKey]];
    self.window.backgroundColor = windowColor ?: [NSColor colorWithRed:0.27 green:0.34 blue:0.49 alpha:1.0];

    NSString *lastWindowFrame = [NSUserDefaults.standardUserDefaults objectForKey:MainWindowFrameKey];
    if (lastWindowFrame) {
        NSRect frame = NSRectFromString(lastWindowFrame);
        [self.window setFrame:frame display:NO];
    }

    __weak typeof(self) weakSelf = self;
    self.titleBarSettingObserver = [KVOBlockObserver observeUserDefaultsKey:ShouldUseTransparentTitleBarKey block:^(BOOL transparent) {
        weakSelf.window.titlebarAppearsTransparent = transparent;
        weakSelf.window.titleVisibility = transparent ? NSWindowTitleHidden : NSWindowTitleVisible;
    }];

    self.titleObserver = [KVOBlockObserver observe:self keyPath:@"contentViewController.webView.title" block:^(id  _Nonnull value) {
        if ([value isKindOfClass:NSString.class]) {
            weakSelf.window.title = value;
        }
    }];
}

- (BOOL)windowShouldClose:(NSWindow *)sender {
    [NSApp hide:sender];
    return NO;
}

- (void)windowDidResize:(NSNotification *)notification {
    if (self.windowLoaded) {
        [NSUserDefaults.standardUserDefaults setObject:NSStringFromRect(self.window.frame) forKey:MainWindowFrameKey];
    }
}

@end

File: Fastmate/MainWindowController.swift
import Cocoa
import Combine
import WebKit

class MainWindowController: NSWindowController, NSWindowDelegate {
    private var subscriptions = Set<AnyCancellable>()

    override func windowDidLoad() {
        super.windowDidLoad()
        setupSubscribers()

        // Fixes that we can't trust that the main window exists in applicationDidFinishLaunching:.
        // Here we always know that this content view controller will be the main web view controller,
        // so inform the app delegate
        let appDelegate = NSApplication.shared.delegate as? AppDelegate
        appDelegate?.mainWebViewController = contentViewController as? WebViewController
    }

    func setupSubscribers() {
        guard let webViewController = contentViewController as? WebViewController, let window = window else {
            return
        }

        webViewController.webView.publisher(for: \.title)
            .replaceNil(with: "Fastmate")
            .assign(to: \.title, on: window)
            .store(in: &subscriptions)

        UserDefaults.standard.publisher(for: \.mainWindowFrame)
            .filter { $0 != NSRect.zero }
            .first()
            .sink { window.setFrame($0, display: false) }
            .store(in: &subscriptions)

        UserDefaults.standard.publisher(for: \.shouldUseTransparentTitleBar)
            .sink {
                window.titlebarAppearsTransparent = $0
                window.titleVisibility = $0 ? .hidden : .visible
            }
            .store(in: &subscriptions)

        NotificationCenter.default.publisher(for: NSWindow.didResizeNotification, object: window)
            .debounce(for: .seconds(0.5), scheduler: DispatchQueue.main)
            .compactMap { ($0.object as? NSWindow)?.frame }
            .assign(to: \.mainWindowFrame, on: UserDefaults.standard)
            .store(in: &subscriptions)

        UserDefaults.standard.publisher(for: \.lastUsedWindowColor)
            .assign(to: \.backgroundColor, on: window)
            .store(in: &subscriptions)
    }

    func windowShouldClose(_ sender: NSWindow) -> Bool {
        NSApp.hide(sender)
        return false
    }
}

